libs/designer/src/lib/core/state/undoRedo/undoRedoTypes.ts

```diff
@@ -0,0 +1,62 @@
+import { updateNodeConnection } from '../../actions/bjsworkflow/connections';
+import type { RootState } from '../../store';
+import {
+  addEdgeFromRunAfter,
+  addImplicitForeachNode,
+  addNode,
+  addSwitchCase,
+  deleteNode,
+  deleteSwitchCase,
+  moveNode,
+  pasteNode,
+  pasteScopeNode,
+  removeEdgeFromRunAfter,
+  updateRunAfter,
+} from '../workflow/workflowSlice';
+
+export interface StateHistory {
+  past: string[];
+  future: string[];
+}
+
+// Omitted slices: designerView, designerOptions, dev, undoRedo
+export type UndoRedoPartialRootState = Pick<
+  RootState,
+  | 'connections'
+  | 'customCode'
+  | 'operations'
+  | 'panel'
+  | 'panelV2'
+  | 'settings'
+  | 'staticResults'
+  | 'tokens'
+  | 'workflow'
+  | 'workflowParameters'
+>;
+
+export const undoableActionTypes = [
+  addNode,
+  moveNode,
+  deleteNode,
+  addSwitchCase,
+  deleteSwitchCase,
+  addImplicitForeachNode,
```

Comment:
Let's remove `addImplicitForeachNode` from this list and manually store a state at the end of `addForeachToNode` in `libs\designer\src\lib\core\utils\loops.ts`